pkg/controller/storagecluster/kms_resources.go

```diff
@@ -0,0 +1,78 @@
+package storagecluster
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	ocsv1 "github.com/openshift/ocs-operator/pkg/apis/ocs/v1"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+const (
+	// KMSConfigMapName is the name configmap which has KMS config details
+	KMSConfigMapName = "ocs-kms-connection-details"
+	// KMSTokenSecretName is the name of the secret which has KMS token details
+	KMSTokenSecretName = "ocs-kms-token"
+	// KMSProviderKey is the key in config map to get the KMS provider name
+	KMSProviderKey = "KMS_PROVIDER"
+	// VaultKMSProvider a constant to represent 'vault' KMS provider
+	VaultKMSProvider = "vault"
+)
+
+func (r *ReconcileStorageCluster) addKmsToCephCluster(instance *ocsv1.StorageCluster, cephCluster *cephv1.CephCluster, reqLogger logr.Logger) error {
+	// if 'KMS' is not enabled, return silently
+	if !instance.Spec.Encryption.KeyManagementService.Enable {
+		return nil
+	}
+	// if provided CephCluster is nil, raise an error
+	if cephCluster == nil {
+		return fmt.Errorf("Empty CephCluster provided")
+	}
+	kmsConfigMap := corev1.ConfigMap{}
+	err := r.client.Get(context.TODO(),
+		types.NamespacedName{
+			Name:      KMSConfigMapName,
+			Namespace: instance.ObjectMeta.Namespace,
+		},
+		&kmsConfigMap,
+	)
+	if err != nil {
+		reqLogger.Error(err, fmt.Sprintf("%s. Unable to get %q config map",
+			"Adding KMS details to cephcluster failed", KMSConfigMapName))
+		return err
+	}
+	if err := isKMSProviderReachable(kmsConfigMap.Data); err != nil {
+		reqLogger.Error(err, "Unable to reach KMS provider address")
+		return err
+	}
+	cephCluster.Spec.Security.KeyManagementService.ConnectionDetails = kmsConfigMap.Data
+	cephCluster.Spec.Security.KeyManagementService.TokenSecretName = KMSTokenSecretName
+	return nil
+}
+
+// isKMSProviderReachable function checks whether the provided address is reachable or not.
+// This function won't validate any other cases and only returns an error if the provided
+// KMS provider address is not reachable. All other validations will be done on rook side.
+func isKMSProviderReachable(kmsConfigMapData map[string]string) error {
+	if kmsConfigMapData == nil {
+		return fmt.Errorf("Please provide a valid configmap data as argument.")
+	}
+	kmsProviderName, ok := kmsConfigMapData[KMSProviderKey]
+	if !ok {
+		return nil
+	}
+	// the provider's address key should be in the format, ${PROVIDER_NAME_IN_CAPS}_ADDR.
+	// For example, for 'vault' KMS provider, the address key will be, 'VAULT_ADDR'
+	kmsProviderAddressKey := fmt.Sprintf("%s_ADDR", strings.ToUpper(kmsProviderName))
```

Comment:
We have 0 guarantee that other KMS implementation will have a `"%s_ADDR"` format so I would `switch case` on `KMSProviderKey` and if `VaultKMSProvider` we use `VAULT_ADDR`.

Reply:
I was trying to avoid any dependency between a new KMS provider addition and code change (at least in OCS-Operator code changes). If we use a conditional approach (adding `if` or `switch` according to the KMS provider), as you mentioned, for any new KMS provider we will have to make code change in the operator.

Another approach (inorder to avoid code changes) is to make the address key (in the configmap data) agnostic to KMS provider, that means for any provider we use the same key, let say `KMS_ENDPOINT` (or something like that).

Let me know your opinion, I'm ok to make changes either way (conditonal or any of the above way).

Reply:
> I was trying to avoid any dependency between a new KMS provider addition and code change (at least in OCS-Operator code changes). If we use a conditional approach (adding `if` or `switch` according to the KMS provider), as you mentioned, for any new KMS provider we will have to make code change in the operator.

Yes, I understood that.

> 
> Another approach (inorder to avoid code changes) is to make the address key (in the configmap data) agnostic to KMS provider, that means for any provider we use the same key, let say `KMS_ENDPOINT` (or something like that).
> 

If make it agnostic, how do we configure the KMS?
Are you saying we should translate `KMS_ENDPOINT` to the right value but inside the Rook code?

> Let me know your opinion, I'm ok to make changes either way (conditonal or any of the above way).

I still think we should go with the conditional since what we have now will only work for Vault.
Thanks

Reply:
Yes changed it to a map, thus user/developer has to just add the KMS provider and address key into the map and no need to edit the code further.